internal/common/tree.go

```diff
@@ -0,0 +1,201 @@
+package common
+
+import (
+	"slices"
+
+	"sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+)
+
+type DNSTreeNode struct {
+	Name     string
+	Children []*DNSTreeNode
+	DataSets []DNSTreeNodeData
+}
+
+type DNSTreeNodeData struct {
+	RecordType       string
+	SetIdentifier    string
+	RecordTTL        endpoint.TTL
+	Labels           endpoint.Labels
+	ProviderSpecific endpoint.ProviderSpecific
+	Targets          []string
+}
+
+// RemoveNode removes a node from a tree.
+// If the node was the only child of the parent node,
+// the parent will be removed as well unless the parent is a root node
+func (n *DNSTreeNode) RemoveNode(deleteNode *DNSTreeNode) {
+	// store indexes of dead branches
+	var deadBranches []int
+	var deadDataSets []string
+
+	if deleteNode == nil {
+		return
+	}
+
+	for i, node := range n.Children {
+		if node.Name == deleteNode.Name {
+			n.Children = append(n.Children[:i], n.Children[i+1:]...)
+
+			// we removed child, but need to clean up data sets
+			for j, dataset := range n.DataSets {
```

Comment:
so this looks fine but I am a bit confused as to why the data set isn't just attached to the node

Reply:
We can have more datasets than nodes. For example, we can have 3 `klb.testdomain.com` records that target two (`eu.klb.testdomain.com` and `us.klb.testdomain.com`) and by having 3 datasets we know to create 3 enpoints of `klb.tesdomain.com`. 

We could "push" datasets to the child (this will solve the need to search on removal) but we will end up with an intricate logic for converting them to endpoints: for each node, we will need to check each child for the number of datasets. And it doesn't look right to store in a child the data you need to convert the parent into the endpoint :thinking: 
here is what i mean: 
```
endpoints := []externaldns.Endpoint{}

for _, child := range node.Children {

    ToEndpoints(child, endpoints)

    for _, dataSet := child.Datasets {
        endpoints = append(endpoints, externaldnsEnpoints{
            DNSName: node.Name
            Targets: dataSet.Targets
            .....
        }
    }
}
```
or something similar. Will have a Node that has dataset for it's parent.

Reply:
eh perhaps it is something we can do beyond this initial impl would rather get this in for now

Reply:
it was suggested to use topology-like approach :thinking:  we will definitely revisit it